src/sse/sse_radix4.rs

```diff
@@ -0,0 +1,518 @@
+        // copy the data into the spectrum vector, split the copying up into chunks to make it more cache friendly
+        let mut num_chunks = signal.len() / 8192;
+        if num_chunks == 0 {
+            num_chunks = 1;
+        } else if num_chunks > self.base_len {
+            num_chunks = self.base_len;
+        }
+        for n in 0..num_chunks {
+            prepare_radix4(
+                signal.len(),
+                self.base_len,
+                signal,
+                spectrum,
+                1,
+                n,
+                num_chunks,
+            );
+        }
+
+        // Base-level FFTs
+        match &self.base_fft {
+            Sse64Butterfly::Len1(bf) => bf.perform_fft_butterfly_multi(spectrum).unwrap(),
+            Sse64Butterfly::Len2(bf) => bf.perform_fft_butterfly_multi(spectrum).unwrap(),
+            Sse64Butterfly::Len4(bf) => bf.perform_fft_butterfly_multi(spectrum).unwrap(),
+            Sse64Butterfly::Len8(bf) => bf.perform_fft_butterfly_multi(spectrum).unwrap(),
+            Sse64Butterfly::Len16(bf) => bf.perform_fft_butterfly_multi(spectrum).unwrap(),
+            Sse64Butterfly::Len32(bf) => bf.perform_fft_butterfly_multi(spectrum).unwrap(),
+        }
+
+        // cross-FFTs
+        let mut current_size = self.base_len * 4;
+        let mut layer_twiddles: &[__m128d] = &self.twiddles;
+
+        while current_size <= signal.len() {
+            let num_rows = signal.len() / current_size;
+
+            for i in 0..num_rows {
+                butterfly_4_64(
+                    &mut spectrum[i * current_size..],
+                    layer_twiddles,
+                    current_size / 4,
+                    &self.bf4,
+                )
+            }
+
+            //skip past all the twiddle factors used in this layer
+            let twiddle_offset = (current_size * 3) / 4;
+            layer_twiddles = &layer_twiddles[twiddle_offset..];
+
+            current_size *= 4;
+        }
+    }
+}
+boilerplate_fft_sse_oop!(Sse64Radix4, |this: &Sse64Radix4<_>| this.len);
+
+fn prepare_radix4<T: FftNum>(
+    size: usize,
+    base_len: usize,
+    signal: &[Complex<T>],
+    spectrum: &mut [Complex<T>],
+    stride: usize,
+    chunk: usize,
+    nbr_chunks: usize,
+) {
+    if size == (4 * base_len) {
+        do_radix4_shuffle(size, signal, spectrum, stride, chunk, nbr_chunks);
+    } else if size == base_len {
+        unsafe {
+            for i in (chunk * base_len / nbr_chunks)..((chunk + 1) * base_len / nbr_chunks) {
+                *spectrum.get_unchecked_mut(i) = *signal.get_unchecked(i * stride);
+            }
+        }
+    } else {
+        for i in 0..4 {
+            prepare_radix4(
+                size / 4,
+                base_len,
+                &signal[i * stride..],
+                &mut spectrum[i * (size / 4)..],
+                stride * 4,
+                chunk,
+                nbr_chunks,
+            );
+        }
+    }
+}
+
+fn do_radix4_shuffle<T: FftNum>(
```

Comment:
Interesting. How much does this improve performance? Since this doesn't involve SSE at all, could we port this to scalar radix4?

Reply:
I didn't measure it directly (yet), but you can see it if you compare the first and second power-of-two comparisons I posted above. Looking at the f32 one, in the first it starts dropping off after 2^12, and in the second it keeps the speed up until after 2^17.
This should help in the scalar radix4 also yes. I also found that similar tricks work on the transpose in mixedradix (still an early WIP).

Reply:
btw as a bonus I think this might help the estimating planner, when the radix4 behaves better at long lengths..